database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala

```diff
@@ -358,4 +362,51 @@ class MetadataSlickDatabase(originalDatabaseConfig: Config)
       labelAndKeyLabelValues, labelOrKeyLabelValues, excludeLabelAndValues, excludeLabelOrValues, submissionTimestampOption, startTimestampOption, endTimestampOption, metadataArchiveStatus, includeSubworkflows)
     runTransaction(action)
   }
+
+  override def deleteNonLabelMetadataForWorkflow(rootWorkflowId: String)(implicit ec: ExecutionContext): Future[Int] = {
+
+    val delete: DBIO[Int] = checkDeletionPreconditions(rootWorkflowId) flatMap { checkResult: Checked[Unit] =>
+      checkResult match {
+        case Right(_) =>
+          dataAccess.metadataEntriesWithoutLabelsForRootWorkflowId(rootWorkflowId).delete
+        case Left(errors: NonEmptyList[String]) =>
+          DBIO.failed(new Exception(errors.toList.mkString(",")))
+      }
+    }
+
+    runTransaction(delete)
+  }
+
+  private def checkDeletionPreconditions(rootWorkflowId: String)(implicit ec: ExecutionContext): DBIO[Checked[Unit]] = {
+    dataAccess.workflowMetadataSummaryEntriesForWorkflowExecutionUuid(rootWorkflowId).result.headOption map {
+      case Some(workflow: WorkflowMetadataSummaryEntry) =>
+        (isTerminalCheck(workflow), isRootWorkflowCheck(workflow)) mapN {
+          case (_, _) => ()
+        }
+      case None =>
+        s"""[Carbonite metadata deletion] Failed with non-existent root workflow "$rootWorkflowId"""".invalidNelCheck
+    }
+  }
+
+  private def isTerminalCheck(summaryEntry: WorkflowMetadataSummaryEntry): Checked[Unit] = {
+    import cromwell.core.WorkflowState
+
+    summaryEntry.workflowStatus match {
+      case Some(status) =>
+        if (WorkflowState.WorkflowStateValues.filter(_.isTerminal).map(_.toString).contains(status))
```

Comment:
For example, the terminal workflow statuses should be passed in as a `statusesOkToDelete: Iterable[String]` that can then be matched in the SQL / slick.